File: iOS/Crisis Connect/Features/TrustDossier/TrustDossierAPI.swift
import FirebaseAuth
import Foundation

enum TrustDossierAPIError: LocalizedError, Equatable {
    case signedOut
    case invalidEndpoint
    case invalidResponse
    case rejected(status: Int, code: String)

    var errorDescription: String? {
        switch self {
        case .signedOut: return NSLocalizedString("Sign in with an authorized institutional account.", comment: "")
        case .invalidEndpoint: return NSLocalizedString("The secure dossier service address is invalid.", comment: "")
        case .invalidResponse: return NSLocalizedString("The secure dossier service returned an invalid response.", comment: "")
        case .rejected(_, let code): return code.replacingOccurrences(of: "-", with: " ")
        }
    }
}

struct TrustDossierActor: Codable, Equatable, Sendable { let uid: String; let name: String }

struct TrustDossierSource: Codable, Equatable, Sendable {
    let kind: String
    let systemKey: String?
    let integrationId: String?
    let externalId: String?
    let externalVersion: String?
}

struct TrustDossierPolicyDecision: Codable, Equatable, Sendable {
    let status: String
    let policyId: String?
    let policyPackId: String?
    let policyPackContentSha256: String?
    let jurisdiction: String
    let signatureRequirement: String
    let organizationSeal: String
    let signerRoles: [String]
    let deliveryReceipt: String
}

struct TrustDossierRecordPlan: Codable, Equatable, Sendable {
    let filePlanCode: String?
    let retentionClass: String
    let retentionUntil: String?
    let legalHold: Bool
}

struct TrustDossierComponentView: Codable, Identifiable, Equatable, Sendable {
    let id: String
    let fileName: String
    let mediaType: String
    let bytes: Int
    let sha256: String
}

struct TrustDossierSummary: Codable, Identifiable, Equatable, Sendable {
    var id: String { dossierId }
    let dossierId: String
    let title: String
    let description: String
    let purpose: String
    let classification: String
    let state: String
    let revision: Int
    let source: TrustDossierSource
    let policy: TrustDossierPolicyDecision
    let recordPlan: TrustDossierRecordPlan
    let components: [TrustDossierComponentView]
    let manifestSha256: String?
    let updatedAtMillis: Int64
}

struct TrustDossierPolicyPackContent: Codable, Equatable, Sendable {
    let signatureRequirement: String
    let organizationSeal: String
    let signerRoles: [String]
    let deliveryReceipt: String
    let retentionClass: String
    let retentionDays: Int
    let filePlanCode: String?
}

struct TrustDossierPolicyPack: Codable, Identifiable, Equatable, Sendable {
    var id: String { packId }
    let packId: String
    let revision: Int
    let status: String
    let name: String
    let jurisdiction: String
    let purpose: String
    let effectiveFrom: String
    let effectiveUntil: String?
    let content: TrustDossierPolicyPackContent
    let contentSha256: String
    let createdBy: TrustDossierActor
}

private struct DossierListEnvelope: Decodable { let dossiers: [TrustDossierSummary] }
private struct DossierEnvelope: Decodable { let dossier: TrustDossierSummary }
private struct PolicyPackListEnvelope: Decodable { let policyPacks: [TrustDossierPolicyPack] }
private struct APIErrorEnvelope: Decodable { let error: String? }

actor TrustDossierAPI {
    static let shared = TrustDossierAPI()
    private let session: URLSession
    private let decoder = JSONDecoder()

    init(session: URLSession = .shared) { self.session = session }

    func list() async throws -> ([TrustDossierSummary], [TrustDossierPolicyPack]) {
        async let dossiers: DossierListEnvelope = request(path: "/api/dashboard/dossiers?limit=60")
        async let packs: PolicyPackListEnvelope = request(path: "/api/dashboard/dossier-policy-packs")
        return try await (dossiers.dossiers, packs.policyPacks)
    }

    func create(title: String, description: String, purpose: String, classification: String,
                jurisdiction: String, filePlanCode: String?) async throws -> TrustDossierSummary {
        let recordPlan: [String: Any] = [
            "filePlanCode": filePlanCode ?? NSNull(),
            "retentionClass": "policy_pending",
        ]
        let body: [String: Any] = ["dossier": [
            "title": title, "description": description, "purpose": purpose,
            "classification": classification, "jurisdiction": jurisdiction,
            "source": ["kind": "standalone", "systemKey": NSNull(), "integrationId": NSNull(),
                       "externalId": NSNull(), "externalVersion": NSNull()],
            "recordPlan": recordPlan,
        ]]
        let envelope: DossierEnvelope = try await request(path: "/api/dashboard/dossiers", method: "POST",
            body: body, idempotencyKey: operationID())
        return envelope.dossier
    }

    func applyPolicy(pack: TrustDossierPolicyPack, to dossier: TrustDossierSummary) async throws -> TrustDossierSummary {
        let retentionUntil = ISO8601DateFormatter().string(from: Date().addingTimeInterval(TimeInterval(pack.content.retentionDays) * 86_400))
        let body: [String: Any] = ["action": "accept_policy", "expectedRevision": dossier.revision,
            "policyPackId": pack.packId, "policyId": pack.packId,
            "signatureRequirement": pack.content.signatureRequirement, "organizationSeal": pack.content.organizationSeal,
            "signerRoles": pack.content.signerRoles, "deliveryReceipt": pack.content.deliveryReceipt,
            "retentionClass": pack.content.retentionClass, "retentionUntil": retentionUntil,
            "filePlanCode": pack.content.filePlanCode ?? NSNull()]
        return try await mutate(dossier: dossier, body: body)
    }

    func freeze(_ dossier: TrustDossierSummary) async throws -> TrustDossierSummary {
        try await mutate(dossier: dossier, body: ["action": "freeze", "expectedRevision": dossier.revision])
    }

    func upload(data: Data, fileName: String, mediaType: String, to dossier: TrustDossierSummary) async throws -> TrustDossierSummary {
        guard data.count <= 10 * 1_024 * 1_024 else { throw TrustDossierAPIError.rejected(status: 413, code: "file-too-large") }
        let boundary = "CrisisConnect-(UUID().uuidString)"
        var body = Data()
        func field(_ name: String, _ value: String) {
            body.append("--\(boundary)\r\nContent-Disposition: form-data; name=\"\(name)\"\r\n\r\n\(value)\r\n".data(using: .utf8)!)
        }
        field("operationId", operationID()); field("expectedRevision", String(dossier.revision))
        body.append("--\(boundary)\r\nContent-Disposition: form-data; name=\"document\"; filename=\"\(safeFileName(fileName))\"\r\nContent-Type: \(mediaType)\r\n\r\n".data(using: .utf8)!)
        body.append(data); body.append("\r\n--\(boundary)--\r\n".data(using: .utf8)!)
        let envelope: DossierEnvelope = try await request(path: "/api/dashboard/dossiers/\(encodedPath(dossier.dossierId))/components",
            method: "POST", rawBody: body, contentType: "multipart/form-data; boundary=\(boundary)")
        return envelope.dossier
    }

    private func mutate(dossier: TrustDossierSummary, body: [String: Any]) async throws -> TrustDossierSummary {
        let envelope: DossierEnvelope = try await request(path: "/api/dashboard/dossiers/\(encodedPath(dossier.dossierId))",
            method: "POST", body: body, idempotencyKey: operationID())
        return envelope.dossier
    }

    private func request<T: Decodable>(path: String, method: String = "GET", body: [String: Any]? = nil,
                                       rawBody: Data? = nil, contentType: String = "application/json; charset=utf-8",
                                       idempotencyKey: String? = nil) async throws -> T {
        guard let user = Auth.auth().currentUser, !user.isAnonymous else { throw TrustDossierAPIError.signedOut }
        let token = try await user.getIDToken()
        var request = try makeRequest(path: path, method: method, token: token, body: body, rawBody: rawBody,
                                      contentType: contentType, idempotencyKey: idempotencyKey)
        var (data, response) = try await session.data(for: request)
        if (response as? HTTPURLResponse)?.statusCode == 401 {
            request.setValue("Bearer \(try await user.getIDToken(forcingRefresh: true))", forHTTPHeaderField: "Authorization")
            (data, response) = try await session.data(for: request)
        }
        guard let http = response as? HTTPURLResponse else { throw TrustDossierAPIError.invalidResponse }
        guard (200...299).contains(http.statusCode) else {
            let code = (try? decoder.decode(APIErrorEnvelope.self, from: data).error) ?? "request-rejected"
            throw TrustDossierAPIError.rejected(status: http.statusCode, code: code)
        }
        guard let value = try? decoder.decode(T.self, from: data) else { throw TrustDossierAPIError.invalidResponse }
        return value
    }

    private func makeRequest(path: String, method: String, token: String, body: [String: Any]?, rawBody: Data?,
                             contentType: String, idempotencyKey: String?) throws -> URLRequest {
        guard let url = endpoint(path: path) else { throw TrustDossierAPIError.invalidEndpoint }
        var request = URLRequest(url: url); request.httpMethod = method; request.timeoutInterval = 45
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        if method != "GET" { request.setValue(contentType, forHTTPHeaderField: "Content-Type") }
        if let idempotencyKey { request.setValue(idempotencyKey, forHTTPHeaderField: "x-cc-idempotency-key") }
        request.httpBody = rawBody ?? (body.map { try? JSONSerialization.data(withJSONObject: $0) } ?? nil)
        return request
    }

    private func endpoint(path: String) -> URL? {
        let configured = (Bundle.main.object(forInfoDictionaryKey: "CRISIS_CONNECT_WEB_URL") as? String)?
            .trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        let base = configured.isEmpty ? "https://crisisconnect.network" : configured
        guard var components = URLComponents(string: base), components.scheme == "https", components.user == nil,
              components.password == nil, let route = URLComponents(string: path) else { return nil }
        components.path = components.path.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        components.path = "/" + [components.path, route.path.trimmingCharacters(in: CharacterSet(charactersIn: "/"))]
            .filter { !$0.isEmpty }.joined(separator: "/")
        components.queryItems = route.queryItems; components.fragment = nil
        return components.url
    }

    private func operationID() -> String { "ios:\(UUID().uuidString)" }
    private func encodedPath(_ value: String) -> String { value.addingPercentEncoding(withAllowedCharacters: .urlPathAllowed) ?? "invalid" }
    private func safeFileName(_ value: String) -> String {
        let filtered = value.unicodeScalars.map { scalar -> Character in scalar.value < 32 || scalar.value == 127 ? "_" : Character(String(scalar)) }
        return String(filtered).replacingOccurrences(of: "\"", with: "_").replacingOccurrences(of: "\\", with: "_").prefix(180).description
    }
}

File: Android/app/src/main/java/com/auralis/crisisconnect/dossier/TrustDossierApi.kt
package com.auralis.crisisconnect.dossier

import com.auralis.crisisconnect.BuildConfig
import com.auralis.crisisconnect.network.PinnedOkHttpClient
import com.google.firebase.auth.FirebaseAuth
import java.io.IOException
import java.time.Instant
import java.time.temporal.ChronoUnit
import java.util.UUID
import java.util.concurrent.TimeUnit
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.tasks.await
import kotlinx.coroutines.withContext
import okhttp3.HttpUrl
import okhttp3.HttpUrl.Companion.toHttpUrlOrNull
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.MultipartBody
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONArray
import org.json.JSONObject

data class DossierComponent(
    val id: String,
    val fileName: String,
    val mediaType: String,
    val bytes: Long,
    val sha256: String,
)

data class DossierPolicy(
    val status: String,
    val jurisdiction: String,
    val policyId: String?,
    val signatureRequirement: String,
    val deliveryReceipt: String,
)

data class TrustDossier(
    val dossierId: String,
    val title: String,
    val description: String,
    val purpose: String,
    val classification: String,
    val state: String,
    val revision: Int,
    val policy: DossierPolicy,
    val retentionClass: String,
    val components: List<DossierComponent>,
    val manifestSha256: String?,
)

data class PolicyPackContent(
    val signatureRequirement: String,
    val organizationSeal: String,
    val signerRoles: List<String>,
    val deliveryReceipt: String,
    val retentionClass: String,
    val retentionDays: Int,
    val filePlanCode: String?,
)

data class TrustDossierPolicyPack(
    val packId: String,
    val status: String,
    val name: String,
    val jurisdiction: String,
    val purpose: String,
    val effectiveFrom: Instant,
    val effectiveUntil: Instant?,
    val content: PolicyPackContent,
)

class TrustDossierApiException(val status: Int?, message: String) : IOException(message)

class TrustDossierApi(
    baseUrl: String = BuildConfig.TRUST_DOSSIER_BASE_URL.trim().trimEnd('/').ifBlank { DEFAULT_BASE_URL },
    private val auth: FirebaseAuth = FirebaseAuth.getInstance(),
    client: OkHttpClient = PinnedOkHttpClient.newClient(),
) {
    private val base: HttpUrl = baseUrl.toHttpUrlOrNull()
        ?.takeIf { it.isHttps && it.username.isEmpty() && it.password.isEmpty() }
        ?: throw IllegalArgumentException("Secure dossier service URL must be HTTPS")
    private val client = client.newBuilder()
        .connectTimeout(15, TimeUnit.SECONDS)
        .readTimeout(45, TimeUnit.SECONDS)
        .writeTimeout(45, TimeUnit.SECONDS)
        .build()

    suspend fun list(): Pair<List<TrustDossier>, List<TrustDossierPolicyPack>> =
        listDossiers() to listPolicyPacks()

    suspend fun create(
        title: String,
        description: String,
        purpose: String,
        classification: String,
        jurisdiction: String,
        filePlanCode: String?,
    ): TrustDossier {
        val dossier = JSONObject()
            .put("title", title.trim())
            .put("description", description.trim())
            .put("purpose", purpose)
            .put("classification", classification)
            .put("jurisdiction", jurisdiction.trim().uppercase())
            .put("source", JSONObject().put("kind", "standalone")
                .put("systemKey", JSONObject.NULL).put("integrationId", JSONObject.NULL)
                .put("externalId", JSONObject.NULL).put("externalVersion", JSONObject.NULL))
            .put("recordPlan", JSONObject().put("filePlanCode", filePlanCode ?: JSONObject.NULL)
                .put("retentionClass", "policy_pending"))
        return parseDossier(postJson(endpoint("api", "dashboard", "dossiers"),
            JSONObject().put("dossier", dossier), operationId()).getJSONObject("dossier"))
    }

    suspend fun applyPolicy(dossier: TrustDossier, pack: TrustDossierPolicyPack): TrustDossier {
        val retentionUntil = Instant.now().plus(pack.content.retentionDays.toLong(), ChronoUnit.DAYS).toString()
        val body = JSONObject()
            .put("action", "accept_policy")
            .put("expectedRevision", dossier.revision)
            .put("policyPackId", pack.packId)
            .put("policyId", pack.packId)
            .put("signatureRequirement", pack.content.signatureRequirement)
            .put("organizationSeal", pack.content.organizationSeal)
            .put("signerRoles", JSONArray(pack.content.signerRoles))
            .put("deliveryReceipt", pack.content.deliveryReceipt)
            .put("retentionClass", pack.content.retentionClass)
            .put("retentionUntil", retentionUntil)
            .put("filePlanCode", pack.content.filePlanCode ?: JSONObject.NULL)
        return mutate(dossier, body)
    }

    suspend fun freeze(dossier: TrustDossier): TrustDossier = mutate(
        dossier,
        JSONObject().put("action", "freeze").put("expectedRevision", dossier.revision),
    )

    suspend fun upload(
        dossier: TrustDossier,
        bytes: ByteArray,
        fileName: String,
        mediaType: String,
    ): TrustDossier {
        require(bytes.size <= MAX_UPLOAD_BYTES) { "file-too-large" }
        require(mediaType in ALLOWED_MEDIA_TYPES) { "unsupported-media-type" }
        val safeName = fileName.replace(Regex("[\\p{Cc}\\\\\"]"), "_").take(180)
        val multipart = MultipartBody.Builder().setType(MultipartBody.FORM)
            .addFormDataPart("operationId", operationId())
            .addFormDataPart("expectedRevision", dossier.revision.toString())
            .addFormDataPart("document", safeName, bytes.toRequestBody(mediaType.toMediaType()))
            .build()
        val response = authenticated { token ->
            Request.Builder().url(endpoint("api", "dashboard", "dossiers", dossier.dossierId, "components"))
                .header("Authorization", "Bearer $token")
                .header("Accept", "application/json")
                .post(multipart)
                .build()
        }
        return parseDossier(response.getJSONObject("dossier"))
    }

    private suspend fun listDossiers(): List<TrustDossier> {
        val url = endpoint("api", "dashboard", "dossiers").newBuilder()
            .addQueryParameter("limit", "60").build()
        val root = authenticated { token -> Request.Builder().url(url)
            .header("Authorization", "Bearer $token").header("Accept", "application/json").get().build() }
        return root.getJSONArray("dossiers").objects().map(::parseDossier)
    }

    private suspend fun listPolicyPacks(): List<TrustDossierPolicyPack> {
        val root = authenticated { token -> Request.Builder()
            .url(endpoint("api", "dashboard", "dossier-policy-packs"))
            .header("Authorization", "Bearer $token").header("Accept", "application/json").get().build() }
        return root.getJSONArray("policyPacks").objects().map(::parsePolicyPack)
    }

    private suspend fun mutate(dossier: TrustDossier, body: JSONObject): TrustDossier =
        parseDossier(postJson(endpoint("api", "dashboard", "dossiers", dossier.dossierId),
            body, operationId()).getJSONObject("dossier"))

    private suspend fun postJson(url: HttpUrl, body: JSONObject, idempotencyKey: String): JSONObject =
        authenticated { token -> Request.Builder().url(url)
            .header("Authorization", "Bearer $token")
            .header("Accept", "application/json")
            .header("x-cc-idempotency-key", idempotencyKey)
            .post(body.toString().toRequestBody(JSON_MEDIA_TYPE))
            .build() }

    private suspend fun authenticated(build: (String) -> Request): JSONObject {
        val user = auth.currentUser?.takeUnless { it.isAnonymous }
            ?: throw TrustDossierApiException(401, "Sign in with an authorized institutional account")
        suspend fun token(force: Boolean): String = user.getIdToken(force).await().token
            ?.trim()?.takeIf(String::isNotEmpty)
            ?: throw TrustDossierApiException(401, "Could not obtain a secure session")
        return try {
            execute(build(token(false)))
        } catch (rejected: TrustDossierApiException) {
            if (rejected.status != 401) throw rejected
            execute(build(token(true)))
        }
    }

    private suspend fun execute(request: Request): JSONObject = withContext(Dispatchers.IO) {
        client.newCall(request).execute().use { response ->
            val text = response.body?.string().orEmpty()
            val json = runCatching { JSONObject(text) }.getOrElse { JSONObject() }
            if (!response.isSuccessful) {
                throw TrustDossierApiException(response.code,
                    json.optString("error").ifBlank { "Secure dossier request rejected (${response.code})" })
            }
            json
        }
    }

    private fun endpoint(vararg path: String): HttpUrl = base.newBuilder().apply {
        path.forEach(::addPathSegment)
    }.build()

    private fun parseDossier(json: JSONObject): TrustDossier {
        val policy = json.getJSONObject("policy")
        val recordPlan = json.getJSONObject("recordPlan")
        return TrustDossier(
            dossierId = json.getString("dossierId"),
            title = json.getString("title"),
            description = json.optString("description"),
            purpose = json.getString("purpose"),
            classification = json.getString("classification"),
            state = json.getString("state"),
            revision = json.getInt("revision"),
            policy = DossierPolicy(
                status = policy.getString("status"),
                jurisdiction = policy.getString("jurisdiction"),
                policyId = policy.nullableString("policyId"),
                signatureRequirement = policy.getString("signatureRequirement"),
                deliveryReceipt = policy.getString("deliveryReceipt"),
            ),
            retentionClass = recordPlan.getString("retentionClass"),
            components = json.getJSONArray("components").objects().map { component ->
                DossierComponent(component.getString("id"), component.getString("fileName"),
                    component.getString("mediaType"), component.getLong("bytes"), component.getString("sha256"))
            },
            manifestSha256 = json.nullableString("manifestSha256"),
        )
    }

    private fun parsePolicyPack(json: JSONObject): TrustDossierPolicyPack {
        val content = json.getJSONObject("content")
        return TrustDossierPolicyPack(
            packId = json.getString("packId"), status = json.getString("status"),
            name = json.getString("name"), jurisdiction = json.getString("jurisdiction"),
            purpose = json.getString("purpose"), effectiveFrom = Instant.parse(json.getString("effectiveFrom")),
            effectiveUntil = json.nullableString("effectiveUntil")?.let(Instant::parse),
            content = PolicyPackContent(
                signatureRequirement = content.getString("signatureRequirement"),
                organizationSeal = content.getString("organizationSeal"),
                signerRoles = content.getJSONArray("signerRoles").strings(),
                deliveryReceipt = content.getString("deliveryReceipt"),
                retentionClass = content.getString("retentionClass"),
                retentionDays = content.getInt("retentionDays"),
                filePlanCode = content.nullableString("filePlanCode"),
            ),
        )
    }

    private fun operationId(): String = "android:${UUID.randomUUID()}"

    companion object {
        const val MAX_UPLOAD_BYTES = 10 * 1024 * 1024
        val ALLOWED_MEDIA_TYPES = setOf("application/pdf", "image/png", "image/jpeg")
        private const val DEFAULT_BASE_URL = "https://crisisconnect.network"
        private val JSON_MEDIA_TYPE = "application/json; charset=utf-8".toMediaType()
    }
}

private fun JSONObject.nullableString(name: String): String? =
    if (isNull(name)) null else optString(name).takeIf(String::isNotBlank)

private fun JSONArray.objects(): List<JSONObject> = (0 until length()).mapNotNull(::optJSONObject)
private fun JSONArray.strings(): List<String> = (0 until length()).mapNotNull { optString(it).takeIf(String::isNotBlank) }
